Open Judge System/Web/OJS.Web/Areas/Administration/Controllers/ExamGroupsController.cs

```diff
@@ -0,0 +1,286 @@
+                return this.GridOperation(request, null);
+            }
+
+            if (!this.UserHasContestRights(examGroup.ContestId.Value))
+            {
+                this.ModelState.AddModelError(string.Empty, GeneralResource.No_privileges_message);
+                return this.GridOperation(request, null);
+            }
+
+            examGroup.Users.Add(user);
+            this.examGroupsData.Update(examGroup);
+
+            var result = new DetailModelType
+            {
+                UserId = user.Id,
+                Username = user.UserName,
+                FirstName = user.UserSettings.FirstName,
+                LastName = user.UserSettings.LastName,
+                Email = user.Email
+            };
+
+            return this.Json(new[] { result }.ToDataSourceResult(request), JsonRequestBehavior.AllowGet);
+        }
+
+        public JsonResult GetAvailableUsers(string userFilter)
+        {
+            var users = this.usersData.GetAll().Take(DefaultUsersTakeCount);
+
+            if (!string.IsNullOrWhiteSpace(userFilter))
+            {
+                users = this.usersData
+                    .GetAll()
+                    .Where(u => u.UserName.ToLower().Contains(userFilter.ToLower()))
+                    .Take(DefaultUsersTakeCount);
+            }
+
+            var result = users
+                .Select(u => new SelectListItem
+                {
+                    Text = u.UserName,
+                    Value = u.Id
+                })
+                .ToList();
+
+            return this.Json(result, JsonRequestBehavior.AllowGet);
+        }
+
+        public JsonResult GetAvailableContests(string contestFilter)
+        {
+            var contests = this.contestsData
+                .GetAll()
+                .OrderByDescending(c => c.CreatedOn);
+
+            if (!this.User.IsAdmin() && this.User.IsLecturer())
+            {
+                contests = contests
+                    .Where(c =>
+                        c.Lecturers.Any(l => l.LecturerId == this.UserProfile.Id) ||
+                        c.Category.Lecturers.Any(l => l.LecturerId == this.UserProfile.Id))
+                    .OrderByDescending(c => c.CreatedOn);
+            }
+
+            if (!string.IsNullOrWhiteSpace(contestFilter))
+            {
+                contests = contests
+                    .Where(c => c.Name.Contains(contestFilter))
+                    .Take(DefaultContestsToTake)
+                    .OrderByDescending(c => c.CreatedOn);
+            }
+
+            var result = contests
+                .Select(c => new
+                {
+                    c.Name,
+                    c.Id
+                })
+                .ToList();
+
+            return this.Json(result, JsonRequestBehavior.AllowGet);
+        }
+
+        public override string GetEntityKeyName() => this.GetEntityKeyNameByType(typeof(ExamGroup));
+
+        private ExamGroup GetByIdAsNoTracking(int id) =>
```

Comment:
I know the rationale behind using `AsNoTracking()` is that we don't want to get tangled up in the data vs. service entity trackers, but is there a more general approach we could take as to eliminate the `data` from the base controller?

Reply:
I agree that we need more general approach, but since it is not so straight forward to implement it,  I made a different issue for this problem: https://github.com/SoftUni-Internal/suls-issues/issues/3964